### PR TITLE
[improve][txn] Cleanup how superusers abort txns

### PR DESCRIPTION
**Pull request by michaeljmarshall**
### Motivation

This PR builds on https://github.com/apache/pulsar/pull/19467. When we modify/abort transactions, we need to make sure that authorization is checked for both the proxy and the client.

### Modifications

* Add a second authorization check when `originalPrincipal` is set in the `ServerCnx`.
* Fix a bug where we were not doing a deep copy of the `SubscriptionsList` object. (Tests caught this bug!)

### Verifying this change

Added a new test to cover some of the changes.

### Does this pull request potentially affect one of the following parts:

This is an internal change.

### Documentation

- [x] `doc-not-needed`

### Matching PR in forked repository

PR in forked repository: https://github.com/michaeljmarshall/pulsar/pull/38

### CODE REVIEW

**Review by nicoloboschi — APPROVED** [review]

**Comment by michaeljmarshall** [comment]
@Technoboy- - I just discovered that this PR will fix a regression introduced by https://github.com/apache/pulsar/pull/19467. Once this is ready to go (hopefully today), I will cherry pick it to `branch-2.11`.

**Comment by Technoboy-** [comment]
> @Technoboy- - I just discovered that this PR will fix a regression introduced by #19467. Once this is ready to go (hopefully today), I will cherry pick it to `branch-2.11`.

Ok

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java [hunk 1]
```diff
@@ -2534,32 +2535,39 @@ protected void handleEndTxn(CommandEndTxn command) {
...
+                return isAuthRoleAuthorized;
+            }
         } else {
             return CompletableFuture.completedFuture(true);
         }
     }
 
     private CompletableFuture<Boolean> verifyTxnOwnership(TxnID txnID) {
-        final String checkOwner = getPrincipal();
+        assert ctx.executor().inEventLoop();
```
Comment (congbobo184) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:2558]: why we need this assert?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java [hunk 2]
```diff
@@ -2534,32 +2535,39 @@ protected void handleEndTxn(CommandEndTxn command) {
...
                     if (service.isAuthenticationEnabled() && service.isAuthorizationEnabled()) {
-                        return getBrokerService()
-                                .getAuthorizationService()
-                                .isSuperUser(checkOwner, getAuthenticationData());
+                        return isSuperUser();
                     } else {
                         return CompletableFuture.completedFuture(false);
                     }
-                });
+                }, ctx.executor());
```
Comment (congbobo184) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java:2570]: Why do we need to change the thread? I think `verifyTxnOwnership` will change the thread right?

**Review by congbobo184 — APPROVED** [review]

**Comment by michaeljmarshall** [comment]
I cherry picked to 2.11 and verified that some of the relevant tests pass.

I will cherry pick to 2.10 tomorrow. There are some conflicts I need to work through.

**Pull request merged**